sqlite3.go

```diff
@@ -79,6 +80,16 @@ _sqlite3_exec(sqlite3* db, const char* pcmd, long long* rowid, long long* change
   return rv;
 }
 
+static const char *
+_trim_leading_spaces(const char *str) {
```

Comment:
Remove, trimming leading spaces is not required. (I don't know why the old code did it, but there's no point as SQLite itself ignores them.)

Reply:
I think this table may help illustrate:

`zSql`|`ppStmt`|`pzTail`
---|---|---
`""`|null|`""`
`" "`|null|`""`
`"/*abc*/"`|null|`""`
`"--abc"`|null|`""`
`";"`|null|`""`
`"; VALUES(1)`|non-null|`""`
`"; VALUES(1); VALUES(2)"`|non-null|`" VALUES(2)"`
`" ;;; VALUES(1)"`|non-null|`""`
`" VALUES(1)"`|non-null|`""`
`"VALUES (1) "`|non-null|`""`
`"VALUES (1) /*abc*/"`|non-null|`""`
`"VALUES (1) --abc"`|non-null|`""`
`"VALUES (1); "`|non-null|`" "`
`"VALUES (1); /*abc*/"`|non-null|`" /*abc*/"`
`"VALUES (1); --abc"`|non-null|`" --abc"`
`"VALUES (1); VALUES (2); VALUES (3);"`|non-null|`"  VALUES (2); VALUES (3);"`

In other words, if `sqlite3_prepare_v2` returns a null `sqlite3_stmt` (with `SQLITE_OK`), then it means the input string did not contain any statements at all, but rather only some combination of whitespace, comments, and semicolons. This is supported by [the docs](https://www.sqlite.org/c3ref/prepare.html).
>  If pzTail is not NULL then *pzTail is made to point to the first byte past the end of the first SQL statement in zSql. These routines only compile the first statement in zSql, so *pzTail is left pointing to what remains uncompiled. [...] If the input text contains no SQL (if the input is an empty string or a comment) then *ppStmt is set to NULL.

Consequently, if the `sqlite3_stmt` is null, that implies that `pzTail` is the empty string. So rather than attempting to trim whitespace (which will not account for comments or superfluous semicolons), we should just let `sqlite3_prepare_v2` do its job. If it returns a null `sqlite3_stmt`, then we are definitely done. And as an optimization, at the top of the loop, we could also break if the current string is empty.

Reply:
Thanks and I'll look into this, but the tests do fail if the white space is not trimmed.

Reply:
@rittneje I was able to finally look into this and the issue is deeper than just trimming whitespace. With the current logic (i.e. `master`) **any** query with a trailing comment will result in an infinite loop.

This commit https://github.com/charlievieth/go-sqlite3/commit/6bc917b7c83bba94c7e2fec4aa827447716f35da modifies [`TestQueryer`](https://github.com/charlievieth/go-sqlite3/blob/6bc917b7c83bba94c7e2fec4aa827447716f35da/sqlite3_test.go#L1093-L1145) to reproduce the error.

<details>
<summary>patch</summary>

```patch
--- a/sqlite3_test.go
+++ b/sqlite3_test.go
@@ -1115,7 +1115,7 @@ func TestQueryer(t *testing.T) {
 		t.Error("Failed to call db.Exec:", err)
 	}
 	rows, err := db.Query(`
-		select id from foo order by id;
+		select id from foo order by id; -- #1 harmless comment
 	`)
 	if err != nil {
 		t.Error("Failed to call db.Query:", err)
@@ -1128,10 +1128,13 @@ func TestQueryer(t *testing.T) {
 		if err != nil {
 			t.Error("Failed to db.Query:", err)
 		}
-		if id != n + 1 {
+		if id != n+1 {
 			t.Error("Failed to db.Query: not matched results")
 		}
 		n = n + 1
+		if n > 3 {
+			t.Fatal("Too many errors:", n)
+		}
 	}
 	if err := rows.Err(); err != nil {
 		t.Errorf("Post-scan failed: %v\n", err)
```
</details>

Fixing this is likely outside the scope of this PR and I'm not sure I'll have the cycles to address it, but will let you know if that changes. Additionally, let me know if I should create an issue for this.

The issue appears to be due to the code not checking if `sqlite3_prepare_v2` sets its `sqlite3_stmt**` argument to NULL because the input text contains no SQL, which can occur if there are comments after a valid SQL statement (see: the [`sqlite3_prepare_v2` docs](https://www.sqlite.org/c3ref/prepare.html)).

A possible solution would be to put more of this logic into the returned `SQLiteRows` struct and let it work through the provided arguments and statements (this would also allow for multiple SQL statements with args in a single query).

Below is a simple example using the `sqlite3` CLI to show that leading/trailing comments are allowed around a SQL QUERY statement.
```sh
#!/bin/sh

set -eu

DB="$(mktemp)"   # test database
TMP1="$(mktemp)" # want
TMP2="$(mktemp)" # got

trap 'rm -f -- ${DB} ${TMP1} ${TMP2}' EXIT

# init/seed table
echo '
drop table if exists foo;
create table foo (id integer);
insert into foo(id) values(1);
insert into foo(id) values(2);
insert into foo(id) values(3);' | sqlite3 "${DB}"

# desired output
echo '1
2
3' >"${TMP1}"

# test "SELECT" statement with leading/trailing comments
echo '
-- FOO
select id from foo order by id;
-- BAR
    /* BAZ */
' | sqlite3 "${DB}" >"${TMP2}"

if ! diff --color "${TMP1}" "${TMP2}"; then
    echo 'FAIL'
    exit 1
fi

echo 'PASS'

```

Reply:
Additionally, with the current logic, if two SELECT statements are provided then the returned Rows are only for the second statement ("select 1; select 2" returns only 2).

I believe this is also the root cause of: https://github.com/mattn/go-sqlite3/issues/950

Reply:
@rittneje just pushed https://github.com/mattn/go-sqlite3/pull/1133/commits/5880fdcffe33877a099908984a130554a533b3f0 which addresses the above 2 bugs. This might be breaking change since in the case of `SELECT 1; SELECT 2;` an error is returned instead of `SELECT 2;` being the query executed.